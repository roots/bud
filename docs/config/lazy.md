## bud.lazy

Enable experimental lazy compilation. Disabled by default.

Read more about webpack's [lazy compilation feature experiment](https://webpack.js.org/configuration/experiments/#experimentslazycompilation)

## Usage

```js
bud.lazy();
```
