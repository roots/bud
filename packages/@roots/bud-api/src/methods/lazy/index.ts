import {Api} from '@roots/bud-framework'

declare module '@roots/bud-framework' {
  interface Framework {
    /**
     * ## bud.lazy
     *
     * Enable experimental lazy compilation. Disabled by default.
     *
     * ### Usage
     *
     * ```js
     * bud.lazy()
     * ```
     *
     * Disable:
     *
     * ```js
     * bud.lazy(false)
     * ```
     */
    lazy: Api.Lazy
  }

  namespace Api {
    type Lazy = (enabled?: boolean) => Framework
  }
}

const lazy: Api.Lazy = function (enabled = true) {
  this.hooks.on(
    'build/experiments/lazyCompilation',
    () => enabled,
  )

  return this
}

export {lazy}
