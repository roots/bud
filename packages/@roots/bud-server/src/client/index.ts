const {overlay} = require('./ErrorOverlay')
const {indicator} = require('./Indicator')

const parsePayload = payload => {
  return {
    complete: payload?.action == 'built',
    pending: payload?.action == 'building',
    hasWarnings: payload?.warnings?.length > 0,
    hasErrors: payload?.errors?.length > 0,
  }
}

const getServerData = async () => {
  const res = await fetch('/__roots/config.json')
  const server = await res.json()

  return server
}

const run = async () => {
  const server = await getServerData()
  const {
    subscribeAll,
    useCustomOverlay,
  } = require(`webpack-hot-middleware/client?quiet=false`)

  if (server.browser.overlay) useCustomOverlay(overlay)

  if (server.browser.indicator) indicator.init()

  subscribeAll(payload => {
    const {hasWarnings, hasErrors, pending, complete} =
      parsePayload(payload)

    if (payload.action == 'reload') window.location.reload()

    if (server.browser.indicator)
      indicator.update({
        payload,
        complete,
        pending,
        hasWarnings,
        hasErrors,
      })
  })
}

run()
