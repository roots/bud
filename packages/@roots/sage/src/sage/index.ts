import {bud, Bud} from '@roots/bud'
import * as babel from '@roots/bud-babel'
import * as esbuild from '@roots/bud-esbuild'
import * as postcss from '@roots/bud-postcss'
import * as react from '@roots/bud-react'
import * as tailwindcss from '@roots/bud-tailwindcss'
import * as typescript from '@roots/bud-typescript'
import * as eslint from '@roots/bud-eslint'
import * as prettier from '@roots/bud-prettier'
import * as stylelint from '@roots/bud-stylelint'
import * as entrypoints from '@roots/bud-entrypoints'
import * as dependencies from '@roots/bud-wordpress-dependencies'
import * as externals from '@roots/bud-wordpress-externals'
import * as manifests from '@roots/bud-wordpress-manifests'

/**
 * Bud preset: @roots/sage
 */
export type Sage = Bud
export const sage: Sage = (sage =>
  sage

    /**
     * Artifacts/cache store
     *
     * Set to Acorn standard location
     */
    .storage('storage/bud')

    /**
     * Src path
     */
    .srcPath('resources')

    /**
     * Dist path
     */
    .distPath('public')

    /**
     * Public path
     */
    .when(
      ({env}: Bud) => env.has('APP_PUBLIC_PATH'),
      ({publicPath, env}: Bud) =>
        publicPath(env.get('APP_PUBLIC_PATH')),
    )

    /**
     * Set CSS urls to be relative
     */
    .config({css: {relativeUrls: true}})

    /**
     * ESBuild doesn't support HMR. it is purely a transpiler.
     *
     * - snowpack & vite each have their own HMR solutions.
     * - snowpack provides their solution as an
     *   [ESM spec proposal](https://github.com/snowpackjs/esm-hmr)
     * - [Related article from Dan Abramov](https://medium.com/@dan_abramov/hot-reloading-in-react-1140438583bf)
     *
     * Losing hmr in development is not worth the theoretical
     * lost time in dev, for most users.
     *
     * Thus, Sage uses esbuild in production, and babel/hmr in development.
     */
    .when(
      ({isDevelopment}) => isDevelopment,
      ({use}: Sage) => use(typescript).use(babel).use(react),
      ({use}: Sage) => use(esbuild).esbuild.jsx(),
    )

    /**
     * General extensions
     */
    .use([
      /**
       * Style transpilation
       */
      postcss,
      tailwindcss,

      /**
       * Linting
       */
      eslint,
      prettier,
      stylelint,

      /**
       * Manifests
       */
      entrypoints,
      dependencies,
      externals,
      manifests,
    ])

    /**
     * Sage webpack aliases
     */
    .alias({
      '@fonts': 'fonts',
      '@images': 'images',
      '@scripts': 'scripts',
      '@styles': 'styles',
    })

    /**
     * Provide
     */
    .provide({
      jquery: ['$', 'jQuery'],
    })

    /**
     * Additional options
     */
    .when(
      ({isProduction}) => isProduction,

      /**
       * Production
       */
      (sage: Sage) => {
        sage.minify().hash().vendor().runtime('single')
      },

      /**
       * Development
       */
      ({env, proxy, dev, devtool}) => {
        dev({
          host: env.has('APP_HOST')
            ? env.get('APP_HOST')
            : 'localhost',
          port: env.has('APP_PORT') ? env.get('APP_PORT') : 3000,
        })

        proxy({
          host: env.has('APP_PROXY_HOST')
            ? env.get('APP_PROXY_HOST')
            : 'localhost',
          port: env.has('APP_PROXY_PORT')
            ? env.get('APP_PROXY_PORT')
            : 8000,
        })

        devtool()
      },
    )
    .html(false))(Object.assign(bud, {name: 'sage'}))
