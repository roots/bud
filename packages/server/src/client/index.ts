import {Configuration} from 'webpack'
import {ansiColors, overlayStyles} from './styles'

interface InjectionProps {
  entrypoints: Configuration['entry']
}

/**
 * Hot client/server module dependencies
 */
const client = require.resolve('webpack-hot-middleware/client')
const params = `path=/__webpack_hmr`
const colors = encodeURIComponent(JSON.stringify(ansiColors))
const styles = encodeURIComponent(JSON.stringify(overlayStyles))

/**
 * Inject webpack entrypoints with client HMR handling script(s).
 */
export type InjectClient = (
  props: InjectionProps,
) => Configuration['entry']

export const injectClient: InjectClient = ({entrypoints}) => {
  const toInject = [
    `${client}?${params}&ansiColors=${colors}&overlayStyles=${styles}`,
  ]

  const prepend = (entry: unknown) => {
    if (typeof entry === 'function') {
      return () => Promise.resolve(entry()).then(prepend)
    }

    if (typeof entry === 'object' && !Array.isArray(entry)) {
      const entryClone = {}

      Object.keys(entry).forEach(key => {
        entryClone[key] = toInject.concat(entry[key])
      })

      return entryClone
    }
  }

  return prepend(entrypoints)
}
