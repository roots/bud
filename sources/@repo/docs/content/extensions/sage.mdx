---
title: '@roots/sage'
description: 'Sage starter theme extension'
sidebar_label: 'ðŸŸ¢ @roots/sage'
---

import Tabs from '@theme/Tabs'
import TabItem from '@theme/TabItem'
import {Install} from '@site/src/docs/Install'
import ThemeJson from '@site/../../@roots/sage/docs/theme.json.mdx'
import TailwindColors from '@site/../../@roots/sage/docs/tailwind.mdx'
import Eslint from '@site/../../@roots/sage/docs/eslint.mdx'
import Stylelint from '@site/../../@roots/sage/docs/stylelint.mdx'
import Sass from '@site/../../@roots/sage/docs/with-sass.mdx'

[@roots/sage](/extensions/sage) is a preset configuration for the [Sage](https://github.com/roots/sage) WordPress starter theme.

## Installation

<Install packages="@roots/sage" />

## Generate `theme.json`

<ThemeJson />

### Defining `theme.json` colors from `tailwind.config.js`

<TailwindColors />

## Linting

### Linting with `eslint`

<Eslint />

### Linting with `stylelint`

<Stylelint />

## Usage with Sass

<Sass />
