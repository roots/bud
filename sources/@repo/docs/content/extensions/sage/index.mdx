---
title: '@roots/sage'
description: 'Sage starter theme extension'
sidebar_label: '@roots/sage'
hide_title: true
hide_table_of_contents: true
---

import {Hero} from '@site/src/components/hero'

import IncludedExtensions from '@site/../../@roots/sage/docs/01-included-extensions.md'

<Hero
  title="@roots/sage"
  subtitle="Preset configuration for the Sage WordPress starter theme"
/>

## Installation

```npm2yarn
npm install @roots/sage --save-dev
```

## Usage

The [Sage docs](https://docs.roots.io/sage/10.x/compiling-assets) cover everything in more detail than can be found here.

## Included extensions

<IncludedExtensions />

## Acorn compatibility

[roots/sage](https://github.com/roots/sage) is built with the [roots/acorn](https://github.com/roots/acorn) framework. This preset guarantees compatibility with Acorn's API for enqueuing JS and CSS.

If you do not wish to use Acorn to enqueue assets you can disable the Acorn extension:

```ts title=bud.config.ts
export default async (bud: Bud) => {
  bud.sage.acorn.enable(false)
}
```

This will restore the default behavior of including the public path (e.g. `/app/themes/sage/public`) in entrypoints manifest values.
