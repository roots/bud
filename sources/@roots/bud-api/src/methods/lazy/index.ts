import type {Bud} from '@roots/bud-framework'
import type {LazyCompilationOptions} from '@roots/bud-framework'

export type Parameters = [boolean | LazyCompilationOptions]

/**
 * Lazy function interface
 */
export interface lazy {
  (...parameters: Parameters): Bud
}

/**
 * Enables lazy compilation of built assets.
 *
 * @example
 * Enable:
 *
 * ```js
 * bud.lazy()
 * ```
 *
 * @example
 * Explicitly disable:
 *
 * ```js
 * bud.lazy(false)
 * ```
 *
 * @example
 * Explicitly enable:
 *
 * ```js
 * bud.lazy(true)
 * ```
 *
 * @example
 * Use with object options:
 *
 * ```js
 * bud.lazy({
 *   entries: false,
 *   imports: true,
 * })
 * ```
 */
export const lazy: lazy = function (this: Bud, value = true) {
  if (typeof value === `boolean`) {
    this.hooks.on(`build.experiments`, experiments => ({
      ...(experiments ?? {}),
      lazyCompilation: value,
    }))
    return this
  }

  this.hooks.on(`build.experiments`, experiments => {
    if (
      typeof experiments?.lazyCompilation === `boolean` ||
      !experiments?.lazyCompilation
    )
      return {...(experiments ?? {}), lazyCompilation: value}

    experiments.lazyCompilation = {
      ...experiments.lazyCompilation,
      ...value,
    }

    return experiments
  })

  return this
}
