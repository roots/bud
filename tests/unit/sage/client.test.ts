import * as pkg from '@roots/sage'
import * as client from '@roots/sage/client'
import domReady from '@roots/sage/client/dom-ready'
import lazy from '@roots/sage/client/lazy'

describe(`@roots/sage`, () => {
  it(`@roots/sage:client`, () => {
    expect(pkg.client).toBe(client)
  })

  it(`@roots/sage/client`, () => {
    expect(client).toMatchSnapshot()
  })

  it(`@roots/sage/client/dom-ready`, () => {
    expect(client.domReady).toBe(domReady)
  })

  it(`@roots/sage/client/lazy`, () => {
    expect(client.lazy).toBe(lazy)
  })
})
