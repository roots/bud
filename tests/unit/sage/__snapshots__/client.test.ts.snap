// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@roots/sage @roots/sage/client 1`] = `
Object {
  "domReady": [Function],
}
`;
